### ISSUE TITLE
[BUG] Possible to hijack sessions upon portal restart

### ISSUE THREAD

**Issue by TehomCD** [evennia/evennia#2055]
#### Describe the bug
We had a game lock, and the portal was restarted while the server remained running. Upon reconnecting, one player was informed that their session was now being shared by someone else's character. Presumably the other player connected with a session that was now being reused. I'm not entirely sure how this happened. Unfortunately I have no information if the player was connecting via the webclient or telnet.

#### To Reproduce
1. Probably have several people logged in
2. Restart portal while server remains running
3. Try to have more people log in

#### Expected behavior
Sessions to not be reused.

#### Environment, Evennia version, OS etc
Evennia version: 0.9.0 (rev 38eb779)
         OS: posix
         Python: 3.8.1
         Twisted: 19.10.0
         Django: 2.2.9

#### Additional context
Maybe additional cleanup on portal/server resync?

**Comment by strikaco** [comment]
Which multisession mode are you running?

**Comment by trhr** [comment]
> one player was informed that their session was now being shared by someone else's character

Can you elaborate more on this? Are you suggesting that Player A was puppeting Player B's character?

My thoughts are as follows.

On connection at the webclient level, the portal does things like checking the django session cookie in the database (which is handled by django auth). During my recent deep dive into portal-server authentication as part of developing an auto-authenticated protocol, I saw no way for two players to both "auth" through the webclient into one another player's account (short of manipulating their own cookies).

On reconnection at the telnet or other protocol level, there is no "resume" functionality to get someone back into their old session. Login is forced.

**Comment by Griatch** [comment]
@TehomCD Could this be related to https://github.com/evennia/evennia/issues/1315 maybe? That was related to auto-logins, and there have been some auto-login PRs merged recently. Could it be that some recent change re-introduced this error again? We never did fully understand why it happened.

**Comment by Griatch** [comment]
To be honest, it's _really_ unexpected to have the Portal restart while the Server staying up ... why would you want to do that? You lose connections when the Portal drops so why would you want to keep the Server running?

**Comment by trhr** [comment]
The Django auth backend queries its own authenticated session storage anyway, right? That's what both a "portal session" and "server session" ultimately resolve to, the sessid that django generates and writes upon a successful auth. So if the django session backend is the sole source of truth about sessions, that feels like a completely sane and bug-free piece of the puzzle.

**Comment by TehomCD** [comment]
I believe player A was seeing all the output of player B's session - I don't know if they had any control over it or not once the connection had been made.

I don't think that the sessions being used/reused are the same thing as the django sessions backend - my understanding is that they're an Evennia construct inside the sessionshandlers, though there might be some crossover in the webclient, maybe.

As far as the portal being restarted while the server stays up - yeah, I don't think that's something we'd do often. Another staff member thought it was the portal rather than the server which was faulty when our AMP bridge became non-functional. Breaking the connection from either side fixed it, but she thought that it was the portal that was at fault, so she only restarted that and didn't think to restart the server. After the issues popped up, we agreed to restart the server in all cases going forward.

It could very well involve some weirdness in the webclient - I don't know if the players in question were using it. I don't know if we have any custom stuff there anymore, but I might have added something years ago and forgot about it. I'll poke around.

**Comment by Griatch** [comment]
@trhr The Server/PortalSessions are authenticated using Django auth, but they are not Django Sessions in the way I think you mean; the Sessions are Evennia-specific constructs, just as Tehom says. 

@TehomCD The very fact that she could reboot the Portal without the Server also getting rebooted is potentially a bug on its own - what exact operations did she use to do that feat?

If the Portal dies and starts up, the Server will auto-connect to it (due to the nature of being an AMP client to the Portal's AMP server) but if the Portal was coming back from shutdown, this would mean its SessionHandler would be empty.The Server's Sessions should really be dropped at that point since with the Portal down, everyone's connections would be gone too. I wonder if there could be something like the Server re-syncing its Sessions with the Portal and messing up which Sessions are live. The Portal is meant to be the source of truth for which Sessions are around, but maybe the system gets confused in a situation where the Portal dies and comes back but the Server is still around.

**Comment by TehomCD** [comment]
I'm not sure - for all I know she killed the portal process while leaving the server running. I'll need to ask her if she remembers. does `evennia stop portal` still work?

**Comment by Griatch** [comment]
Yes, `evennia stop portal` does work, but that will (should) also kill the Server automatically before allowing the Portal to shut down. The command line will report if the Server was killed or was already stopped.

**Comment by TehomCD** [comment]
Oh, the server must have failed to shut down then. The nature of the bug they were addressing prevented the server from ever shutting down cleanly, it always had to be killed.

**Comment by Griatch** [comment]
Hm, the Portal waits for the Server connect to close before it is allowed to itself shut down. I suppose if the Server is already messed up to the point of the AMP connection being non-functioning, the Portal may find the connection is already down ...

**Comment by trhr** [comment]
> @trhr The Server/PortalSessions are authenticated using Django auth, but they are not Django Sessions in the way I think you mean; the Sessions are Evennia-specific constructs, just as Tehom says.

Got it, no sole source of truth about who is authenticated/logged in currently; two separate systems, with one (the webclient) able to bypass the second via what's essentially a pointer to an integer it expects to be there.

And instead of working within the django sessions backend and storing/killing uuid4 serversessions like one might store/kill an http session (aka a traditional website django auth), we prefer to just clean up the integer whenever the pointer might get misaligned.

**Comment by Griatch** [comment]
@trhr Ok, so if this is what happens, any suggestion for improvements of this mechanism?

**Comment by trhr** [comment]
My understanding is that functionally, this is an (inadvertant) CSRF/session-riding attack orchestrated by a portal restart. Here are my thoughts:

Session-riding prevention usually comes down to not allowing communications down channels which are not explicitly trusted by the authentication backend.

After a server or portal restart, all session pipes at the amp level should become untrusted, and trust should be reestablished via checking all sessions against a sole source of truth, i.e. django sessions.

Accounts connecting to sessions which cannot be automatically authenticated as "the same pipe" should be forced to reauthenticate.

Objects should never be puppeted by untrusted sessions (nor by sessions connected to an account not authorized to puppet that character, etc), and if an object's session loses trust at the django-sessions level, the object should become unpuppeted.

**Comment by Griatch** [comment]
While this seems extremely unlikely to happen, it would be addressed by https://github.com/evennia/evennia/issues/2171
